src/server/journal/journal_slice.cc

```diff
@@ -157,28 +157,35 @@ void JournalSlice::CallOnChange(const JournalItem& item) {
   // CallOnChange is atomic iff JournalSlice::SetFlushMode(false) is called before.
   std::shared_lock lk(cb_mu_);
 
-  const size_t size = change_cb_arr_.size();
-  auto k_v = change_cb_arr_.begin();
+  const size_t size = journal_consumers_arr_.size();
+  auto k_v = journal_consumers_arr_.begin();
   for (size_t i = 0; i < size; ++i) {
-    k_v->second(item, enable_journal_flush_);
+    k_v->second->ConsumeJournalChange(item);
     ++k_v;
   }
+  k_v = journal_consumers_arr_.begin();
+  if (enable_journal_flush_) {
+    for (size_t i = 0; i < size; ++i) {
```

Comment:
move size initialization here and add a comment that you set size because ThrottleIfNeeded can preempt and meanwhile journal_consumers_arr_ can grow. and btw, calling ThrottleIfNeeded on a new consumer is harmless so now you can remove `size` workaround at all, imho